common/workunit/workflow.cpp

```diff
@@ -613,12 +707,672 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+{
+#ifdef TRACE_WORKFLOW
+if(currentWfid)
+    LOG(MCworkflow, "Performing workflow item %u", wfid);
+#endif
+    //timestamp_type startTime = getTimeStampNowValue();
+    //CCycleTimer timer;
+    process->perform(ctx, wfid);
+    //noteTiming(wfid, startTime, timer.elapsedNs());
+}
+bool WorkflowMachine::handleFailureParallel(CCloneWorkflowItem & item, WorkflowException const * e, bool isDep)
+{
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "Workflow item %u failed. Aborting task", item.queryWfid());
+#endif
+    abort = true;
+    item.setAbortFlag(true);
+    item.setException(e);
+    StringBuffer msg;
+    e->errorMessage(msg).append(" (in item ").append(e->queryWfid()).append(")");
+    if(isDep)
+        logctx.logOperatorException(NULL, NULL, 0, "Dependency failure for workflow item %u: %d: %s", item.queryWfid(), e->errorCode(), msg.str());
+    else
+        logctx.logOperatorException(NULL, NULL, 0, "%d: %s", e->errorCode(), msg.str());
+    item.setFailInfo(e->errorCode(), msg.str());
+    switch(item.queryType())
+    {
+    case WFTypeNormal:
+        item.setState(WFStateFail);
+        break;
+    case WFTypeSuccess:
+    case WFTypeFailure:
+        item.setState(WFStateNull);
+        break;
+    case WFTypeRecovery:
+        item.setState(WFStateSkip);
+        break;
+    }
+    unsigned failureWfid = item.queryFailure();
+    if(failureWfid)
+    {
+        pendingContingencies ++;
+        queryWorkflowItem(failureWfid).setAbortFlag(true);
+        addToItemQueue(failureWfid);
+        return true;
+    }
+    return false;
+}
+CCloneWorkflowItem &WorkflowMachine::queryWorkflowItem(unsigned wfid)
+{
+    if(wfid <= workflow->count())
+    {
+        return static_cast<CCloneWorkflowItem&>(workflow->queryWfid(wfid));;
+    }
+    else
+    {
+        unsigned index = wfid - workflow->count() - 1;
+        if(index >= logicalWorkflow.size())
+            throwUnexpected();
+        return static_cast<CCloneWorkflowItem&>(*logicalWorkflow[index].get());
+    }
+}
+void WorkflowMachine::initialiseItemQueue()
+{
+    //Items without any predecessors can be executed by a thread
+    //all runtime workflow items should have predecessors or be contingencies
+    for(int i = 1; i <= workflow->count(); i++)
+    {
+        CCloneWorkflowItem & cur = static_cast<CCloneWorkflowItem&>(workflow->queryWfid(i));
+
+        if(cur.queryNumPredecessors() == 0)
+            itemQueue.push(i);
+    }
+    itemQueueSem.signal(itemQueue.size());
+}
+bool WorkflowMachine::checkIfDone()
+{
+    if((pendingContingencies == 0) && (rootReached))
+    {
+        if( (branchCount == 0) || failureClauseOld)
+        {
+        #ifdef TRACE_WORKFLOW
+                LOG(MCworkflow, "workflow done");
+        #endif
+            done = true;
+            itemQueueSem.signal(numThreads);
+            return true;
+        }
+    }
+    return false;
+}
+void WorkflowMachine::processWfItems()
+{
+    unsigned currentWfid = 0;
```

Comment:
this can now be moved to the front() call

Reply:
please could you clarify the suggestion?